dali/pipeline/executor/source_info_propagation.cc

```diff
@@ -0,0 +1,150 @@
+// Copyright (c) 2022, NVIDIA CORPORATION & AFFILIATES. All rights reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include "dali/pipeline/executor/source_info_propagation.h"
+
+#include <cassert>
+#include <functional>
+#include <string>
+
+#include "dali/pipeline/data/tensor_list.h"
+#include "dali/pipeline/workspace/workspace.h"
+
+namespace dali {
+
+template <typename Backend>
+bool SourceInfoDefined(const TensorList<Backend> &tl) {
+  for (int s = 0; s < tl.num_samples(); s++)
+    if (!tl.GetMeta(s).GetSourceInfo().empty())
+      return true;
+  return false;
+}
+
+inline bool OutputSourceInfoDefined(Workspace &ws) {
+  for (int o = 0; o < ws.NumOutput(); o++) {
+    if (ws.OutputIsType<GPUBackend>(o)) {
+      if (SourceInfoDefined(ws.Output<GPUBackend>(o)))
+        return true;
+    } else {
+      assert(ws.OutputIsType<CPUBackend>(o));
+      if (SourceInfoDefined(ws.Output<CPUBackend>(o)))
+        return true;
+    }
+  }
+  return false;
+}
+
+inline int GetOutputBatchSize(Workspace &ws, int output_idx) {
```

Comment:
Wouldn't this `inline` be harmful in a way? I believe, that if you declare non-static inline function, the compiler would error if there appears another function of the same name and different signature in separate translation unit.

> There may be more than one definition of an inline function in the program as long as each definition appears in a different translation unit and **(for non-static inline functions) all definitions are identical**.

How about removing the `inlines` and putting all cc-only functions in a `namespace`?

Also, in the standard
> the meaning of the keyword inline for functions came to mean "multiple definitions are permitted" rather than "inlining is preferred"

So I believe compiler would ignore optimization anyhow.

Reply:
> Wouldn't this inline be harmful in a way? I believe, that if you declare non-static inline function, the compiler would error if there appears another function of the same name and different signature in separate translation unit

No. You even quoted it below: inline for functions came to mean "multiple definitions are permitted". On a free function `inline` is almost synonymous with `static` (and therefore synonymous with placing the function in an anonymous namespace`).